### PR TITLE
FWCore/Utilities : Changes needed to compile with macOS clang and libc++

### PR DESCRIPTION
**Pull request by gartung**

### CODE REVIEW

File: FWCore/Utilities/src/ObjectWithDict.cc [hunk 1]
```diff
@@ -74,9 +76,13 @@ namespace edm {
     }
 
     if (to.hasBase(from)) { // down cast
+#ifndef _LIBCPP_VERSION
       // use the internal dynamic casting of the compiler (e.g. libstdc++.so)
       void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
```
Comment (Dr15Jones) [inline FWCore/Utilities/src/ObjectWithDict.cc]: @davidlt will this work with the new ABI? If not, is there an equivalent function?

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Utilities/src/ObjectWithDict.cc (hunk 1 above)
Comment (davidlt) [inline FWCore/Utilities/src/ObjectWithDict.cc]: `abi::__dynamic_cast` is part of `libsupc++` in GNU world and part of `libc++abi` in LLVM world. `__dynamic_cast` in `libc++ab` is supported on Linux, Darwin and ARM systems and we don't have `libc++abi` in our LLVM toolchain builds (something that we can add).

ABI wise, I don't expect any issues as "Itanium 64-bit ABI" doesn't change, only C++ Standard Library.

Comment (davidlt) [inline FWCore/Utilities/src/ObjectWithDict.cc]: @gartung could you comment on why you need to disable this part?
From https://libcxxabi.llvm.org/spec.html
`void* __dynamic_cast(const void* __src_ptr, const __class_type_info* __src_type, const __class_type_info* __dst_type, ptrdiff_t __src2dst);`
From GNU:
```
  void*
  __dynamic_cast(const void* __src_ptr, // Starting object.
		 const __class_type_info* __src_type, // Static type of object.
		 const __class_type_info* __dst_type, // Desired target type.
		 ptrdiff_t __src2dst); // How src and dst are related.
```
In GNU it's under `namespace abi = __cxxabiv1;` and same applies for LLVM implementation: `namespace abi = __cxxabiv1;`

Comment (gartung) [inline FWCore/Utilities/src/ObjectWithDict.cc]: MacOS clang doesn't link libc++abi by default. I can try linking it in and see if the original will compile.

Comment (davidlt) [inline FWCore/Utilities/src/ObjectWithDict.cc]: Interesting, does it link to something else instead? IIUC, the compiler should generate the calls `__dynamic_cast` and those are handled by the library.

https://github.com/llvm-mirror/clang/blob/a0badfbffbee71c2c757d580fc852d2124dadc5a/test/CodeGenCXX/dynamic-cast.cpp
https://github.com/llvm-mirror/clang/blob/fe32c6a33461a8c60e18c0414d4844a47442328a/lib/CodeGen/ItaniumCXXABI.cpp#L1140

Comment (gartung) [inline FWCore/Utilities/src/ObjectWithDict.cc]: /Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-ywgcs6iqugqoum574jaso7syo63lyzg2/CMSSW_9_0_X/src/FWCore/Utilities/src/ObjectWithDict.cc:81:70: error: no type named '__class_type_info' in namespace '__cxxabiv1'
      void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
                                                                ~~~~~^
/Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-ywgcs6iqugqoum574jaso7syo63lyzg2/CMSSW_9_0_X/src/FWCore/Utilities/src/ObjectWithDict.cc:81:87: error: expected '>'
      void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
                                                                                      ^
                                                                                      >
/Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-ywgcs6iqugqoum574jaso7syo63lyzg2/CMSSW_9_0_X/src/FWCore/Utilities/src/ObjectWithDict.cc:81:64: note: to match this '<'
      void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
                                                               ^
/Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-ywgcs6iqugqoum574jaso7syo63lyzg2/CMSSW_9_0_X/src/FWCore/Utilities/src/ObjectWithDict.cc:81:132: error: no type named '__class_type_info' in namespace '__cxxabiv1'
      void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
                                                                                                                              ~~~~~^
/Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-ywgcs6iqugqoum574jaso7syo63lyzg2/CMSSW_9_0_X/src/FWCore/Utilities/src/ObjectWithDict.cc:81:149: error: expected '>'
      void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
                                                                                                                                                    ^
                                                                                                                                                    >
/Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-ywgcs6iqugqoum574jaso7syo63lyzg2/CMSSW_9_0_X/src/FWCore/Utilities/src/ObjectWithDict.cc:81:126: note: to match this '<'
      void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
                                                                                                                             ^
/Users/gartung/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/fwlite-9.0.X-ywgcs6iqugqoum574jaso7syo63lyzg2/CMSSW_9_0_X/src/FWCore/Utilities/src/ObjectWithDict.cc:81:28: error: no member named '__dynamic_cast' in namespace '__cxxabiv1'
      void* address = abi::__dynamic_cast(address_, static_cast<abi::__class_type_info const*>(&from.typeInfo()), static_cast<abi::__class_type_info const*>(&to.typeInfo()), -1);
                      ~~~~~^
5 errors generated.

Comment (davidlt) [inline FWCore/Utilities/src/ObjectWithDict.cc]: On my machine (clean, no homebrew, MacPorts, etc) I have `/usr/lib/libc++abi.dylib`

Comment (gartung) [inline FWCore/Utilities/src/ObjectWithDict.cc]: macOS clang 8.0.0 is equivalent to llvm 3.8 so maybe its not defined.

Comment (gartung) [inline FWCore/Utilities/src/ObjectWithDict.cc]: What header is it defined in?

Comment (davidlt) [inline FWCore/Utilities/src/ObjectWithDict.cc]: Quick look.
It's implemented in `src/private_typeinfo.cpp`
```
 624 extern "C" _LIBCXXABI_FUNC_VIS void *
 625 __dynamic_cast(const void *static_ptr, const __class_type_info *static_type,
 626                const __class_type_info *dst_type,
 627                std::ptrdiff_t src2dst_offset) {
```

It's not available via headers, but the symbols exists and it's exported:
```
0000000000025a33 T ___dynamic_cast
```
This is probably a safeguard to make sure that only compiler calls it (makes sense).

Comment (davidlt) [inline FWCore/Utilities/src/ObjectWithDict.cc]: The only wait for the user to call this is to define extern interface and link to `/usr/lib/libc++abi.dylib`, same what compiler does. Have you studied why this is needed in the first place? Does C++ not provide us something?

Comment (Dr15Jones) [inline FWCore/Utilities/src/ObjectWithDict.cc]: This is needed to handle case of multiple inheritance where the pointer to the base and pointer to the derived are not the same. That is unlikely in this case (which is why in the distant pass we just assumed they were the same).
Unfortunately, standard C++ doesn't offer anything like this.

Comment (davidlt) [inline FWCore/Utilities/src/ObjectWithDict.cc]: This kind of thing should be explained in a big comment next to this line :) The current comment doesn't give enough details.

**Comment by Dr15Jones** [comment]
Is `_LIBCPP_VERSION` strictly a clang macro or might a version of gcc define it as well?

**Comment by davidlt** [comment]
https://sourceforge.net/p/predef/wiki/Libraries/

It's defined by libc++.

**Comment by davidlt** [comment]
Btw, doesn't this change break CMSSW on Clang + libc++ build? You can also leave it as-is, but on macOS + clang + libc++ system you need to define extern symbol for `__dynamic_cast` and that should solve the problem.

**Comment by gartung** [comment]
As far as I know the clang build of CMSSW uses libstdc++ because clang is compiled with gcc.

**Comment by gartung** [comment]
Why don't you test it and find out.

**Comment by Dr15Jones** [comment]
+1

**Comment by davidlange6** [comment]
+1

**Pull request merged**